### PR TITLE
fix(BUX-497): Routes are hardcoded instead of initialized by configured capabilities

### PR DESCRIPTION
**Pull request by chris-4chain**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/go-paymail/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/go-paymail/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

Initially, the main task was to fix the bug that the routes of capabilities were hardcoded, regardless of what was configured.
That led to introduce `CallableCapability` & `StaticCapability` types, which are used to configure capabilities and automatically configure the router.

### TL;DR What has been done:
1. Single point capabilities definition
- two types of capabilities
    - CallableCapability - has its own http endpoint 
    - StaticCapability - it is a parameter which is visible in json response of `/.well-known/bsvalias` (for now, we have only one static capability - `BRFCSenderValidation`)
- With* config functions which previously manipulated the `CapabilitiesPaylod`, now set appropriate boolean flags in `Configuration` struct
- Previously, the `WithCapabilities` config function did not make much sense, because although it was possible to add custom capabilities, the hard-coded router was not able to handle them. 
    - Now, since the router is automatically generated based on the `CallableCapability`'ies, which contain handlers for each capability, it is possible to add custom capabilities which will be automatically configured by the router.
- Tests have been adjusted
- In `server.server_test.go` there was a commented out test, named 'run server'
    - I made it work and developed it a bit further, so that it tests whether the server is running and `/.well-known/` endpoint as well as all discovered capabilities are available
    - I used `net/http/httptest` package to test the server
- I made the `run_server` example work on localhost with additional&optional custom capabilities
- Function `showCapabilities` and method `EnrichCapabilities` have been refactored and adjusted to the new capabilities concept
- Several minor `TODOs` have been resolved
- There was an additional bug I fixed:
    -  when server was listening on some port (let's say 8080), the links in service discovery were pointing to correct domain but without port.

### CODE REVIEW

File: server/config.go [hunk 1]
```diff
@@ -53,42 +59,32 @@ func (c *Configuration) Validate() error {
...
-		return
+		return true
 	}
 
-	// Sanitize the domain (standard)
 	var err error
 	if domain, err = paymail.SanitizeDomain(domain); err != nil {
-		// todo: log the error? This should rarely occur
-		return
+		c.Logger.Warn().Err(err).Msg("failed to sanitize domain")
```
Comment (arkadiuszos4chain) [inline server/config.go:81]: Why warning?

File: server/config.go [hunk 2]
```diff
@@ -115,42 +111,24 @@ func (c *Configuration) AddDomain(domain string) (err error) {
...
-		}
-	}
-	return capabilities
-}
-
 // GenerateServiceURL will create the service URL
-func GenerateServiceURL(prefix, domain, apiVersion, serviceName string) string {
-
-	// Require prefix or domain
+func GenerateServiceURL(prefix, domain, apiVersion, serviceName string) (string, error) {
```
Comment (arkadiuszos4chain) [inline server/config.go:115]: Shouldn't be moved to capabilities.go?

Bonus question- are you sure there are benefits of using builder for such small string? (I'm just curious)

Comment (chris-4chain) [reply server/config.go:115]: > Shouldn't be moved to capabilities.go?

It's used only in `capabilities.go`, but since it's a public function whose functionality extends beyond the scope of capabilities, I decided not to move it from its original location. However, if you would like me to move it, please let me know, as it doesn't make much difference to me

> Bonus question- are you sure there are benefits of using builder for such small string? (I'm just curious)

I made a benchmark. The results shows that the `old` implementation is slightly better when *short* strings are concerned. For example for a string: `https://bux.4chain.space:1234/v1/bsvalias` (made of 41chars) the *old* implementation wins. 

image
image


Another question is whether this function really matters in terms of performance. I've observed that this function must be called with every `/.well-known/bsvalias` request, and it cannot be calculated during the initialization stage due to the arg:host, which is not constant. That's why I decided to refactor it in the first place.

File: server/config.go (hunk 1 above)
Comment (chris-4chain) [reply server/config.go:81]: My initial policy regarding 'warning or error' was as follows:

- Error: if this event occurred due to a bug or incorrect configuration.
- Warning: if this event occurred due to incorrect input data from a user.

Later, I discovered the function p2pDestination in p2p_payment_destination.go. This function is used to check whether a user has provided the correct domain in their paymail. Based on this, I decided to opt for a warning.

However, if you're not convinced, I am willing to change it to Err.

**Review by arkadiuszos4chain — APPROVED** [review]
